### ISSUE TITLE
[BUG] Memory leak in 6.14.1

### ISSUE THREAD

**Issue by lukasnevosad** [taskrabbit/elasticsearch-dump#628]
**Context**

- ElasticDump 6.14.1
- Elasticsearch 7.3.0
- Node.js 12.10.0
- OS X 10.14.5, confirmed in Debian 9 in Docker.

```
elasticdump --input={elasticUrl} --input-index=places --output=export.json.txt --type=data --concurrency=5 --limit=10000
```

**Describe the bug**
After upgrade from 6.10.6 → 6.14.1 elasticdump runs out of available memory after processing a fraction of the index (~ 500MB of about 400GB). Reproduced on 2 systems. Downgrading back to 6.10.6 seems to have solved the issue.

**Current behavior**
Fails with out of memory:
```

<--- Last few GCs --->

[79:0x3bf4190]   420864 ms: Mark-sweep 999.9 (1003.4) -> 999.8 (1004.1) MB, 1388.8 / 0.0 ms  (average mu = 0.097, current mu = 0.001) allocation failure scavenge might not succeed
[79:0x3bf4190]   422283 ms: Mark-sweep 1000.7 (1004.1) -> 1000.6 (1004.9) MB, 1415.2 / 0.0 ms  (average mu = 0.051, current mu = 0.002) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2abd3bb008a1 <JSObject>
    0: builtin exit frame: parse(this=0x2abd3bb1f199 <Object map = 0x159a55f85341>,0x310ddeb00111 <Very long string[3257244]>,0x2abd3bb1f199 <Object map = 0x159a55f85341>)

    1: parse [0x3cd7868f26c9] [/usr/lib/node_modules/elasticdump/lib/jsonparser.js:29] [bytecode=0x29eeb53700c1 offset=68](this=0x3cd7868f2701 <Object map = 0x8f8aec7aa1>,0x310ddeb00111 <Very long string[3257244]>,0x2ae22aecb...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x9d33e0 node::Abort() [node]
 2: 0x9d4596 node::OnFatalError(char const*, char const*) [node]
 3: 0xb32dbe v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xb33139 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xcde455  [node]
 6: 0xcdeae6 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [node]
 7: 0xcea97a v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0xceb885 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xcee298 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [node]
10: 0xcc117e v8::internal::Factory::NewJSObjectFromMap(v8::internal::Handle<v8::internal::Map>, v8::internal::AllocationType, v8::internal::Handle<v8::internal::AllocationSite>) [node]
11: 0xdc8b96 v8::internal::JsonParser<unsigned short>::BuildJsonObject(v8::internal::JsonParser<unsigned short>::JsonContinuation const&, std::vector<v8::internal::JsonProperty, std::allocator<v8::internal::JsonProperty> > const&, v8::internal::Handle<v8::internal::Map>) [node]
12: 0xdc943f v8::internal::JsonParser<unsigned short>::ParseJsonValue() [node]
13: 0xdca33f v8::internal::JsonParser<unsigned short>::ParseJson() [node]
14: 0xbe41b5 v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [node]
15: 0x136d639  [node]
Aborted
```

**Comment by ferronrsmith** [comment]
Is this inside docker ?

Try allocating more memory `NODE_OPTIONS="--max-old-space-size=8192"`

**Comment by lukasnevosad** [comment]
I already tried running the same outside the docker and It fails the same way. When running outside the docker, htop shows node process steadily climbing to the top (ordered by memory), while in 6.10.6 it stays low.

While debugging, I have also found another memory issue on SO (5 days old) https://stackoverflow.com/questions/58338191/elastic-dump-fail-with-java-script-out-pf-memory/58411356#58411356

**Comment by mattful** [comment]
Context
Elasticdump 6.15.1
Elasticsearch 6.2.4
Docker

I ran into the same issue importing from file. While setting NODE_OPTIONS="--max-old-space-size=8192" got past the "JavaScript heap out of memory", the node process steadily consumes memory until it reaches the container limit (4G) and is oom-killed. Less than 2% of the objects are sent before it gets killed.

$ docker run -rm -it -v $PWD/tmp --memory-4g -e 'NODE_OPTIONS="--max-old-space-size=8192" '  taskrabbit/elasticsearch-dump --type=data --fsCompress --input=/tmp/index.json.gz --output=http://elasticsearch:9200/index

The import file index.json.gz is 3G and contains 165M documents. It was created with an older version of elasticdump (w/out --fsCompress option) and compressed using gzip. 

Using different --limit values (10K, 1K and 500) had no notable impact.

**Comment by ferronrsmith** [comment]
Try v6.15.6

**Comment by mattful** [comment]
Confirm fixed in v6.15.7 (haven't tried v6.15.6). No need for NODE_OPTIONS="--max-old-space-size=8192" either. Thanks!
